core/src/apps/zcash/f4jumble.py

```diff
@@ -0,0 +1,58 @@
+"""
+Memory-optimized implementation of F4jumble permutation specified in ZIP-316.
+specification: https://zips.z.cash/zip-0316#jumbling
+reference implementation: https://github.com/zcash/librustzcash/blob/main/components/f4jumble/src/lib.rs
+"""
+
+from micropython import const
+
+from trezor.crypto.hashlib import blake2b
+
+HASH_LENGTH = const(64)
+
+
+def xor(target: memoryview, mask: bytes) -> None:
+    for i in range(len(target)):
+        target[i] ^= mask[i]
+
+
+def G_round(i: int, left: memoryview, right: memoryview) -> None:
+    for j in range((len(right) + HASH_LENGTH - 1) // HASH_LENGTH):
+        mask = blake2b(
+            personal=b"UA_F4Jumble_G" + bytes([i]) + j.to_bytes(2, "little"),
+            data=bytes(left),
+        ).digest()
+        xor(right[j * HASH_LENGTH : (j + 1) * HASH_LENGTH], mask)
+
+
+def H_round(i: int, left: memoryview, right: memoryview) -> None:
+    mask = blake2b(
+        personal=b"UA_F4Jumble_H" + bytes([i, 0, 0]),
+        outlen=len(left),
+        data=bytes(right),
+    ).digest()
+    xor(left, mask)
+
+
+def f4jumble(message: memoryview) -> None:
+    assert len(message) >= 22
```

Comment:
Why 22?

Upstream says the min length is 48 (and max length is 4194368): https://github.com/zcash/librustzcash/blob/37fc28634e811fdf8db9274a0080cff17c6b6a09/components/f4jumble/src/lib.rs#L67-L70

Reply:
You are right.

I started by copy-pasting implementation by daira. Then I found that it's outdated and I forgot to change this.

https://github.com/daira/jumble/blob/main/f4jumble.py#L18

Reply:
https://github.com/trezor/trezor-firmware/pull/2398/commits/8dd8eff3e1e4c1fa0593987a0d3b697bcd5aed0a